src/nvim/fileio.c

```diff
@@ -978,38 +978,28 @@ int readfile(char *fname, char *sfname, linenr_T from, linenr_T lines_to_skip,
           if (read_buf_lnum > from) {
             size = 0;
           } else {
-            int n, ni;
-            long tlen;
-
-            tlen = 0;
+            size_t tlen = 0;
             for (;;) {
               p = (char_u *)ml_get(read_buf_lnum) + read_buf_col;
-              n = (int)STRLEN(p);
-              if ((int)tlen + n + 1 > size) {
+              size_t n = STRLEN(p);
+              if (tlen + n + 1 > (size_t)size) {
                 // Filled up to "size", append partial line.
                 // Change NL to NUL to reverse the effect done
                 // below.
-                n = (int)(size - tlen);
-                for (ni = 0; ni < n; ni++) {
-                  if (p[ni] == NL) {
-                    ptr[tlen++] = NUL;
-                  } else {
-                    ptr[tlen++] = (char)p[ni];
-                  }
-                }
+                n = (size_t)size - tlen;
+                memmove(ptr + tlen, p, n);
+                // NUL is represented as NL; convert
+                memchrsub(ptr + len, NL, NUL, n);
```

Comment:
does memchrsub() cancel the benefit?

Reply:
I did the same thing in diff.c in #20796 and I observed a 15% perf uplift. My guess is that `memmove` and parts of `memchrsub` can be compiled down to simd instructions.

I haven't figured out how to benchmark this block yet.